### PR TITLE
apiserver: add key/value pair to httplog

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
Adding structured key/value pair to httplog.

Use case
```
httplog.AddKeyValue(ctx, "my key", "my value")
``` 

#### Which issue(s) this PR fixes:
Fixes #

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 1]
```diff
@@ -181,6 +182,16 @@ func AddInfof(ctx context.Context, format string, data ...interface{}) {
 	}
 }
 
+func (rl *respLogger) AddKeyValue(key, value interface{}) {
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:185]: note to self: `key` should be string
```
AddKeyValue(key string, value interface{})
```

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 2]
```diff
@@ -181,6 +182,16 @@ func AddInfof(ctx context.Context, format string, data ...interface{}) {
...
+	rl.addedKeyValuePairs = append(rl.addedKeyValuePairs, key, value)
+}
+
+func AddKeyValue(ctx context.Context, key, value interface{}) {
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:189]: `key` should be string
```
AddKeyValue(key string, value interface{})
```

File: staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go [hunk 3]
```diff
@@ -64,7 +65,7 @@ func TrackCompleted(handler http.Handler) http.Handler {
 		latency := completedAt.Sub(fr.startedTimestamp)
 		metrics.RecordFilterLatency(ctx, fr.name, latency)
 		if klog.V(3).Enabled() && latency > minFilterLatencyToLog {
-			httplog.AddInfof(ctx, "%s=%s", fr.name, latency.String())
+			httplog.AddKeyValue(ctx, fmt.Sprintf("fl_%s", fr.name), latency.String())
```
Comment (tkashem) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:68]: @MikeSpreitzer I also took the liberty to:
- use `AddKeyValue` 
- and add a prefix of `fl` to denote filter latency.  so we see something like `fl_priorityandfairness="122.622271ms"`

Let me know if it is okay, otherwise we can make this change in a separate PR

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filterlatency/filterlatency.go:68]: Using AddKeyValue here looks good to me.
Using `fmt.Sprintf` to concatenate two strings looks unnecessarily heavyweight to me.

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go [hunk 4]
```diff
@@ -50,12 +50,13 @@ const respLoggerContextKey respLoggerContextKeyType = iota
...
-	status         int
-	statusStack    string
-	addedInfo      strings.Builder
-	startTime      time.Time
+	hijacked           bool
+	statusRecorded     bool
+	status             int
+	statusStack        string
+	addedInfo          strings.Builder
+	addedKeyValuePairs []interface{}
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:58]: How about limiting the number of memory objects and allocations added by this PR?
We could make the creation of a respLogger initialize addedKeyValuePairs to `make([]interface{}, 0, 20)` (or whatever turns out to be the common case), so that only one memory alloc and one memory object is added (plus the stuff happening in `Log`, where this PR follows the established pattern; the following idea could improve that too, but should probably not be mixed into this PR).  We could even go further an add a `backingKeyValuePairs [20]interface{}` field and initialize addedKeyValuePairs to `backingKeyValuePairs[:0]`, that would add zero memory objects and allocations in the normal case.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:58]: On second thought, initializing `addedKeyValuePairs` to `backingKeyValuePairs[12:]` and then changing the logic in `Log` that allocates `keysAndValues` to logic that writes to `backingKeyValuePairs[:12]` would be nice and easy.

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:58]: How does `backingKeyValuePairs` work given we have a number of concurrent requests?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:58]: In general, in the current code we initialize with nil, and the first append grows it. 
Given that currently we only use it in one place, this seems good enough to me - we can optimize it later.

**Review by wojtek-t — COMMENTED** [review]
/lgtm
/approve

/hold - to give Mike`s a chance to react

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 4 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:58]: Each request gets its own `respLogger`, so each request would have its own `backingKeyValuePairs`.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:58]: Yes, let's get this interface improvement in ASAP.  The optimizations I suggested are isolated in the implementation.

**Review by MikeSpreitzer — APPROVED** [review]
/unhold

File: staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:58]: > Each request gets its own respLogger, so each request would have its own backingKeyValuePairs.

In which case I don't understand the value from having it - I would say that you get the value from it if you could share them between calls by not allocating it all the time.

Anyway - yes, let's merge and improve later.

Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go:58]: @MikeSpreitzer I will open a separate PR with the optimization.

**Comment by MikeSpreitzer** [comment]
BTW, regarding optimizations: see https://gist.github.com/MikeSpreitzer/a32998a8e547d40cc55696cff097d18c for an escape analysis using go1.16.7.  Ugh, value in the slice of interfaces escapes to the heap.  We could maybe cut out half of them by putting the `interface{}("particular string")` in global variables, but ugh.
